### PR TITLE
[produce][spaceship] update bundle capability API and add new bundle capabilities

### LINKED ISSUE #16233: `Sign in with Apple` option for produce command
### Feature Request

Add `Sign in with Apple` option to `Produce`

#### Motivation Behind Feature

I am trying to automate bootstrapping xcode projects for my company. I use `produce` command for generating App IDs and enabling push notification and associated domains feature. But I want to enable `Sign in with Apple` as well.

#### Feature Description

Add `--sign-in-with-apple` to produce's enable_services command.

### LINKED ISSUE #16429: Spaceship cannot set dataProtection, and nils out existing entries if you try
### New Issue Checklist

- [X] Updated fastlane to the latest version
- [X] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [X] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [X] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
Spaceship seems unable to set `data_protection`. Code snippet:

```
app = Spaceship::Portal.app.find(bundleId)
services = app.details.features
puts("Initial Features: " + services.to_s)
service = Spaceship::Portal::AppService.data_protection.until_first_auth
app.update_service(service)
app = Spaceship::Portal.app.find(bundleId)
services = app.details.features
puts("Updated Features: " + services.to_s)
```

Assuming the initial Features show "" for data protection, the updated features will as well, like so:
```
Initial Features: {"push"=>true, "iCloud"=>false, "inAppPurchase"=>true, "gameCenter"=>true, "LPLF93JG7M"=>false, "passbook"=>false, "IAD53UNK2F"=>false, "V66P55NK2I"=>false, "dataProtection"=>"", "SKC3T5S89Y"=>false, "APG3427HIY"=>true, "HK421J6T7P"=>false, "homeKit"=>false, "WC421J6T7P"=>false, "cloudKitVersion"=>2, "AWEQ28MY3E"=>true}
Updated Features: {"push"=>true, "iCloud"=>false, "inAppPurchase"=>true, "gameCenter"=>true, "LPLF93JG7M"=>false, "passbook"=>false, "IAD53UNK2F"=>false, "V66P55NK2I"=>false, "dataProtection"=>"", "SKC3T5S89Y"=>false, "APG3427HIY"=>true, "HK421J6T7P"=>false, "homeKit"=>false, "WC421J6T7P"=>false, "cloudKitVersion"=>2, "AWEQ28MY3E"=>true}
```

Notice `"dataProtection"=>""` in both cases.

When run against an app that _has_ a dataProtection value, attempting an update will turn that value into the empty string.

##### Full script
```
require "spaceship"

appleId = "you@domain.com"
bundleId = "com.domain.portalTest"


def loginOrFail(appleId)

    if appleId.nil?
        puts "You must supply your Apple Developer Portal credentials."
        exit(false)
    end

    begin
        puts("Logging in as " + appleId)
        Spaceship::Portal.login(appleId)

    rescue
        puts "Login has failed. Your username/password may be incorrect."
        exit(false)
    end

    # Just in case the user is a member of multiple teams
    Spaceship::Portal.select_team
    return true
end


begin
    loginOrFail(appleId)
rescue
    abort("Unable to log in.")
end

app = Spaceship::Portal.app.find(bundleId)
if app.nil?
    puts("Could not find an app with bundleId '" + bundleId + "' on the portal.")
    exit(2)
end

services = app.details.features
puts("Initial Features: " + services.to_s)

service = Spaceship::Portal::AppService.data_protection.until_first_auth

puts(service.service_uri.to_s)
puts(service.service_id.to_s)
puts(service.value.to_s)

app.update_service(service)

updatedApp = Spaceship::Portal.app.find(bundleId)
if updatedApp.nil?
    puts("Could not find an app with bundleId '" + bundleId + "' on the portal.")
    exit(2)
end

newServices = updatedApp.details.features
puts("Updated Features: " + newServices.to_s)
```

### Environment

I am specifically scripting _just_ Spaceship, and not using Fastlane on my projects. So `fastlane -env` does nothing in my current environment, other than try to configure a `fastlane` directory for a non-existent project.

If there is an alternate thing I can do to supply info you need, please let me know.  

- `fastlane --version` returns `fastlane 2.146.1`
- `fastlane spaceship --version` returns `spaceship 2.146.1`

**Edit**
I installed fastlane into a different folder so I could get env:

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 10.15.4                                                        |
| Ruby                        | 2.6.3                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.24.2 (Apple Git-127)                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | Mac OS X 10.15.4 (19E287)                                      |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 11.4.1                                                         |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# This file contains the fastlane.tools configuration
# You can find the documentation at https://docs.fastlane.tools
#
# For a list of all available actions, check out
#
#     https://docs.fastlane.tools/actions
#
# For a list of all available plugins, check out
#
#     https://docs.fastlane.tools/plugins/available-plugins
#

# Uncomment the line if you want fastlane to automatically update itself
# update_fastlane

default_platform(:ios)

platform :ios do
  desc "Description of what the lane does"
  lane :custom_lane do
    # add actions here: https://docs.fastlane.tools/actions
  end
end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
# app_identifier("[[APP_IDENTIFIER]]") # The bundle identifier of your app
# apple_id("[[APPLE_ID]]") # Your Apple email address


# For more information about the Appfile, see:
#     https://docs.fastlane.tools/advanced/#appfile

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.146.1 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.4.0        |
| slack-notifier            | 2.3.2        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| public_suffix             | 2.0.5        |
| babosa                    | 1.0.3        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| faraday                   | 0.17.3       |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.13.1       |
| fastimage                 | 2.1.7        |
| gh_inspector              | 1.1.3        |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| dotenv                    | 2.7.5        |
| bundler                   | 2.1.4        |
| jwt                       | 2.1.0        |
| uber                      | 0.1.0        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.36.4       |
| emoji_regex               | 1.0.1        |
| nanaimo                   | 0.2.6        |
| colored2                  | 3.1.2        |
| claide                    | 1.0.3        |
| CFPropertyList            | 3.0.2        |
| atomos                    | 0.1.3        |
| xcodeproj                 | 1.16.0       |
| unicode-display_width     | 1.7.0        |
| tty-screen                | 0.7.1        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| excon                     | 0.73.0       |
| unf_ext                   | 0.0.7.7      |
| json                      | 2.3.0        |
| mini_magick               | 4.10.1       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| multi_json                | 1.14.1       |
| signet                    | 0.14.0       |
| os                        | 1.1.0        |
| memoist                   | 0.16.2       |
| googleauth                | 0.12.0       |
| digest-crc                | 0.5.1        |
| google-cloud-errors       | 1.0.0        |
| google-cloud-env          | 1.3.1        |
| google-cloud-core         | 1.5.0        |
| google-cloud-storage      | 1.26.1       |
| aws-eventstream           | 1.1.0        |
| aws-sigv4                 | 1.1.3        |
| aws-partitions            | 1.310.0      |
| jmespath                  | 1.4.0        |
| aws-sdk-core              | 3.94.1       |
| aws-sdk-kms               | 1.30.0       |
| aws-sdk-s3                | 1.63.1       |
| forwardable               | 1.3.1        |
| logger                    | 1.4.2        |
| stringio                  | 0.1.0        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.1.0        |
| strscan                   | 1.0.3        |
| date                      | 3.0.0        |
| fileutils                 | 1.4.1        |
| io-console                | 0.5.6        |
| zlib                      | 1.1.0        |
| libxml-ruby               | 3.1.0        |
| rexml                     | 3.2.4        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.6.0        |
</details>

*generated on:* **2020-05-07**
</details>

### PR DESCRIPTION
**Pull request by ShyamLad**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Resolves #16429
Resolves #16233

Apple is changing their developer portal APIs and the old undocumented ones Fastlane is using are losing functionality. One functionality that has been lost is the ability to set the bundle capability `data_protection`, as well as the many new bundle capabilities that now exist.

### Description

I have found a new undocumented API that can be utilized to set bundle capabilities. After analyzing the network traffic of Xcode and the Apple Developer Portal, I can confirm that this new API is being used in both of these applications. 

The first set of changes is to swap the API endpoint to the new one. Luckily, the authentication headers/cookies are the same as the old API. There was also a need to create a generalized request-body object in order to set the new bundle capabilities.

The second set of changes is to update the list of bundle capabilities. There were a bunch of new capabilities added over time that were not supported by the old API, as well as some new ones that were added recently as part of WWDC21.

I directly called SpaceShip using one of our tools to ensure that the ability to set bundle capabilities works with the new API and the new bundle capability options.
I also set up a dummy iOS project, added a Gemfile that pointed to the local fastlane build, and successfully enabled and disabled the bundle capabilities.

Added tests to Spaceship to ensure that the request body is being configured correctly for the new API.
All Produce and Spaceship tests pass.

### Testing Steps

```bundle exec fastlane produce enable_services --access-wifi --app-attest --app-group --apple-pay --associated-domains --auto-fill-credential --class-kit --icloud=xcode6_compatible --custom-network-protocol --data-protection=complete --extended-virtual-address-space --game-center=mac --health-kit --hls-interstitial-preview --home-kit --hotspot --in-app-purchase --inter-app-audio --low-latency-hls --managed-associated-domains --maps --multipath --network-extension --nfc-tag-reading --personal-vpn --passbook --push-notification  --siri-kit --system-extension --user-management --vpn-configuration --wallet --wireless-accessory --car-play-audio-app --car-play-messaging-app --car-play-navigation-app --car-play-voip-calling-app --critical-alerts --hotspot-helper --driver-kit --driver-kit-endpoint-security --driver-kit-family-hid-device --driver-kit-family-networking --driver-kit-family-serial --driver-kit-hid-event-service --driver-kit-transport-hid --multitasking-camera-access --sf-universal-link-api --vp9-decoder --music-kit --shazam-kit --communication-notifications --group-activities --health-kit-estimate-recalibration --time-sensitive-notifications --fonts --family-controls --file-provider-testing-mode --sign-in-with-apple```

```bundle exec fastlane produce disable_services --access-wifi --app-attest --app-group --apple-pay --associated-domains --auto-fill-credential --class-kit --icloud --custom-network-protocol --data-protection --extended-virtual-address-space --health-kit --hls-interstitial-preview --home-kit --hotspot --inter-app-audio --low-latency-hls --managed-associated-domains --maps --multipath --network-extension --nfc-tag-reading --personal-vpn --passbook --push-notification --sign-in-with-apple --siri-kit --system-extension --user-management --vpn-configuration --wallet --wireless-accessory --car-play-audio-app --car-play-messaging-app --car-play-navigation-app --car-play-voip-calling-app --critical-alerts --hotspot-helper --driver-kit --driver-kit-endpoint-security --driver-kit-family-hid-device --driver-kit-family-networking --driver-kit-family-serial --driver-kit-hid-event-service --driver-kit-transport-hid --multitasking-camera-access --sf-universal-link-api --vp9-decoder --music-kit --shazam-kit --communication-notifications --group-activities --health-kit-estimate-recalibration --time-sensitive-notifications --fonts --family-controls --file-provider-testing-mode```

```bundle exec fastlane produce enable_services --data-protection=complete```

```bundle exec fastlane produce disable_services --data-protection```

The following lane:

```Ruby

default_platform(:ios)

platform :ios do
  desc "Test to enable services"
  lane :custom_lane do
  end
  lane :test do
    modify_services(
      username: 'XXXX@google.com',
      app_identifier: 'com.google.XXXXXXX',
      services: {
        access_wifi: "on",                      
        app_attest: "on",                       
        app_group: "on",                        
        apple_pay: "on",                        
        associated_domains: "on",               
        auto_fill_credential: "on",             
        car_play_audio_app: "on",               
        car_play_messaging_app: "on",           
        car_play_navigation_app: "on",          
        car_play_voip_calling_app: "on",        
        class_kit: "on",                        
        icloud: "xcode5_compatible",            
        critical_alerts: "on",                  
        custom_network_protocol: "on",          
        data_protection: "complete",            
        extended_virtual_address_space: "on",   
        file_provider_testing_mode: "on",       
        fonts: "on",                            
        game_center: "ios",                     
        health_kit: "on",                       
        hls_interstitial_preview: "on",         
        home_kit: "on",                         
        hotspot: "on",                          
        hotspot_helper: "on",                   
        in_app_purchase: "on",                  
        inter_app_audio: "on",                  
        low_latency_hls: "on",                  
        managed_associated_domains: "on",       
        maps: "on",                             
        multipath: "on",                        
        network_extension: "on",                
        nfc_tag_reading: "on",                  
        passbook: "on",                         
        personal_vpn: "on",                     
        push_notification: "on",                
        sign_in_with_apple: "on",               
        siri_kit: "on",                         
        system_extension: "on",                 
        user_management: "on",                  
        vpn_configuration: "on",                
        wallet: "on",                           
        wireless_accessory: "on",               
        driver_kit: "on",                       
        driver_kit_endpoint_security: "on",     
        driver_kit_family_hid_device: "on",     
        driver_kit_family_networking: "on",     
        driver_kit_family_serial: "on",         
        driver_kit_hid_event_service: "on",     
        driver_kit_transport_hid: "on",         
        multitasking_camera_access: "on",       
        sf_universal_link_api: "on",            
        vp9_decoder: "on",                      
        music_kit: "on",                        
        shazam_kit: "on",                       
        communication_notifications: "on",      
        group_activities: "on",                 
        health_kit_estimate_recalibration: "on",
        time_sensitive_notifications: "on",     
    })
  end
end
```

### CODE REVIEW

**Review by joshdholtz — CHANGES_REQUESTED** [review]
@ShyamLad Thank you so much for making this PR! I was actually holding off on making this change until the App Store Connect API updates were announced and pushed for WWDC21. The changes that you made here actually belong in the `Spaceship::ConnectAPI` module 😇 This module already handles JSON API parsing and has a `BundleID` class already.

I don't think we will be able to accept this PR as it stands right now in its implementation but the functionality of this PR I would like to accept 🙌 If you want to rework this PR to be use everything in the `Spaceship::ConnectAPI` module that would be 🔥 Otherwise I had this on my list to tackle here soon as well 😊 I would love your contribution but I also want to respect your time ❤️

**Comment by ShyamLad** [comment]
Hey @joshdholtz 

Thanks for the quick reply! I can try to rework the changes to use `Spaceship::ConnectAPI`. To clarify what you want: Would it be acceptable if `Spaceship::PortalClient` called `Spaceship::ConnectAPI` directly when updating bundle capabilities?

**Comment by joshdholtz** [comment]
@ShyamLad Thank YOU for a quick reply 😊 That goal her with `Spaceship::PortalClient` is to eventually deprecate it (if possible). So tools should all switch over to using `Spaceship::ConnectAPI` if possible. `Spaceship::ConnectAPI` handles both App Store Connect API Key auth and also Apple ID auth (which filters back into `Spaceship::PortalClient`). The reason this change needs to take place 100%T in `Spaceship::ConnectAPI` client is because we also need to support App Store Connect API Keys with this change 😇

**Comment by ShyamLad** [comment]
@joshdholtz is `Spaceship::Portal::App` to be deprecated as well?

**Comment by joshdholtz** [comment]
@ShyamLad Yup yup! Goal is to not add anything into anything in `Spaceship::Portal` (unless necessary). Anything that gets removed or deprecated _should_ have a new App Store Connect API endpoint that we should be migrating too

**Comment by ShyamLad** [comment]
@joshdholtz ah ok got it! So ideally you would want `Produce::Service` to send its enable/disable requests directly to `Spaceship::ConnectAPI` and cut out `Spaceship::Portal` entirely?

**Comment by joshdholtz** [comment]
@ShyamLad That is correct! Tools like `deliver` and `pilot` use `Spaceship::ConnectAPI` only but `produce` was not converted yet because only a small amount of the APIs were available so it didn't make sense at the time if it was still working 🤷‍♂️ New App Store Connect APIs haven't been officially announced yet for WWDC21 so I was waiting until those were announced to fully make these changes so that we had official doc to go off of 😊

**Comment by ShyamLad** [comment]
@joshdholtz Got it, I think I will try to swap that over! One more question: Does the current api used for creating bundle capabilities in `Spaceship::ConnectAPI` work with Enterprise? The API I found (which is also being used by the Apple Developer Portal and Xcode) is a PACH request to a slightly different path to the same API host and seems to work for Enterprise.

**Comment by ShyamLad** [comment]
Hey @joshdholtz ,

I made some changes from the suggestions you mentioned. ```Produce::Service``` now directly calls ```Spaceship::ConnectAPI``` (no longer dependent on ```Spaceship::Portal```)  and is able to modify the bundle capabilities. One thing to note - I used the new ```PATCH``` request I had mentioned earlier due to two reasons:
1.  The ```PATCH``` api has the ability to delete, create, and modify bundle capabilities
2. The current API used to delete bundle capabilities in ```Spaceship::ConnectAPI``` does not seem to work for the new Additional Services ```MusicKit``` and ```ShazamKit``` for some reason

Let me know what you think!

**Comment by tmaly1980** [comment]
@joshdholtz @ShyamLad Any progress with this? We're needing to implement Sign in With Apple on white labeled apps and not having to toggle the capability manually would be really useful.

**Comment by joshdholtz** [comment]
@tmaly1980 Hi! This requires quite a bit of testing in my side before I can merge this in and I plan on getting time to finish this up tomorrow 💪 This should go out in a release either end of day tomorrow or Monday

File: spaceship/lib/spaceship/connect_api/provisioning/provisioning.rb
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/provisioning/provisioning.rb:97]: These don't seem to be needed to patch a bundle id capability. Thoughts on removing these 😇

**Review by joshdholtz — CHANGES_REQUESTED** [review]
This looks **really** good 😍 I just have one question/suggestion on what is needed for patching the capability. I ran this locally and found that only the `teamId` attribute was needed 🤷‍♂️

File: spaceship/lib/spaceship/connect_api/provisioning/provisioning.rb
Comment (ShyamLad) [reply spaceship/lib/spaceship/connect_api/provisioning/provisioning.rb:97]: Hey @joshdholtz 

Thanks for reviewing! When you tested with only `teamId` and enabled one capability, did the other enabled capabilities get removed? I noticed that Xcode and the Developer Portal were setting the `delete` field in the `permissions` attribute to `true` and would pass in the entire set of enabled capabilities every time something was changed. The reason I kept this field is to ensure that nothing gets deleted if the user only modifies a certain subset of capabilities. If you saw that the default behavior without this attribute is to preserve the already enabled fields, then I can remove this one. Will also remove the other unnecessary entries.

Comment (joshdholtz) [reply spaceship/lib/spaceship/connect_api/provisioning/provisioning.rb:97]: @ShyamLad Yup yup, they still there! Everything still works. Those attributes that the ones on the "bundle id" itself so we don't really want to change of those. We are only interested in changing the relationship 🤷‍♂️

Comment (ShyamLad) [reply spaceship/lib/spaceship/connect_api/provisioning/provisioning.rb:97]: @joshdholtz Perfect! Will remove this, update the tests, and request a review from you when its completed

Comment (ShyamLad) [reply spaceship/lib/spaceship/connect_api/provisioning/provisioning.rb]: fixed!

**Review by joshdholtz — APPROVED** [review]
I think we are good here! Thank you sooooooo much for reworking this to use `Spaceship::ConnectAPI` 😊 I really appreciate this ❤️

**Comment by spherefoundry** [comment]
Hey @ShyamLad while reviewing the changes to `produce/lib/produce/service.rb` I noticed a little typo around line 85:

```
      if options.access_wifi
         UI.message("\tApp Attest")
         bundle_id.update_capability(APP_ATTEST, enabled: on)
       end
```

I guess the condition should be `options.app_attest`? Sadly I'm not able to sign the CLA due to legal reasons. So I can't submit an PR myself 😞 But it would be cool is someone could.

BTW: great job

**Comment by ShyamLad** [comment]
@spherefoundry ahh nice catch! I'll put out a PR to fix this. Thanks for finding this!

**Pull request merged**
